### DISCUSSION TITLE
[Accepted] S3 Backend plugin for Zenoh

### DISCUSSION THREAD

**Discussion by clibequilibrium** [eclipse-zenoh/roadmap#53] [Ideas and new features]
**Hello Zenoh community !**

I would like to start discussion and propose an idea of using S3 Backend as a storage for Zenoh. 
Currently Zenoh supports fs, influxdb, rocksdb, sql but are there any plans to support S3 storage for example [Minio](https://github.com/minio/minio)? 

And do we see any benefit in it ? I am currently using Zenoh in a distributed social platform where I do aim to support around a million of CCU (concurrent users) that will pull persistent data from S3 bucket storage and will communicate with Zenoh peers to receive the replicated state of the game world . I think S3 falls ideally with key value storage as S3 provides path / url protocol and buckets that could be accessed by Zenoh and the users. 

What are your thoughts on that? I will be happy to discuss !

**Comment by JEnoch** [comment]
Thanks for your suggestion!
We agree that a S3 backend would be a nice feature for lot of use cases with Zenoh.
I add it to the roadmap for a future release, and we'll start to work on it very soon.

**Reply by JEnoch** [reply]
**Added to the roadmap as #54**

**Comment by JEnoch** [comment]
# Design discussion

Here is my "naive" proposal for a first implementation of a S3 backend. Comments are welcome!

## Mapping S3 concepts to Zenoh

### Connectivity to a S3 instance
The simplest is to configure the connectivity to a S3 instance (endpoint, region, credentials...) once for the Backend. This allow to check the connectivity at Backend creation, and not for each of its Storage creation.
If a Zenoh router is required to manage Storages connected to different S3 instances, the user could still create several S3 Backends with distinct configuration for each.

### S3 Bucket (_"A container for stored objects"_)
Make 1 Zenoh storage to use several Buckets seems complex, as it would require to define a mapping between the key expressions to store and the Bucket to store each.
While configuring the Bucket at Backend level (i.e. for all the Zenoh storages it creates) seems too restrictive.
The good balance is to configure 1 Bucket per Zenoh storage.

### Operations mapping
This looks straightforward:
 - a Zenoh `PUT` is mapped to a S3 `PutObject`
 - a Zenoh `DELETE` is mapped to a S3 `DeleteObject`
 - a Zenoh `GET` is mapped to a S3 `GetObject`

### S3 Tags (_"You can add tags to new objects when you upload them. Each tag is a key-value pair"_)
There is no such concept in Zenoh. This first version of S3 plugin won't support tags.

### Zenoh DataInfo
The DataInfo (timestamp, encoding, kind) needs also to be stored.
 - the Zenoh encoding will map to the S3 Object's Content-Type (with is a standard MIME type)
 - the timestamp and kind can be mapped as [user-defined metadata](https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingMetadata.html#UserMetadata)
